File: ncarg2d/src/libncarg/plotchar/pcmpxy.f
C
C $Id: pcmpxy.f,v 1.12 2008-07-27 00:17:20 haley Exp $
C
C                Copyright (C)  2000
C        University Corporation for Atmospheric Research
C                All Rights Reserved
C
C The use of this Software is governed by a License Agreement.
C
      SUBROUTINE PCMPXY (IMAP,XINP,YINP,XOTP,YOTP)
C
C By default, PLCHHQ and PLCHMQ draw characters using X/Y coordinates
C in the fractional system and position the whole character string
C relative to the point (CUFX(XPOS),CUFY(YPOS)).  Setting the flag
C 'MAP' non-zero causes them instead to draw characters using X/Y
C coordinates in an arbitrary X/Y coordinate system and to position
C the whole character string relative to the point (XPOS,YPOS); just
C before drawing, PCMPXY is called to map these coordinates into the
C current user system.
C
C The user of PLOTCHAR may replace this routine as desired to transform
C the characters being drawn.  By default, 'MAP' = 1 selects the EZMAP
C transformations, 'MAP' = 2 selects a polar coordinate transformation,
C 'MAP' = 3 selects a perspective transformation, and 'MAP' = 4 selects
C a scheme appropriate for labelling points on the globe, as described
C in the next paragraph.
C
C When 'MAP' = 4, in all user calls to PLCHHQ and PLCHMQ the values
C of the arguments XPOS, YPOS, and ANGD should be zero.  The argument
C SIZE specifies the desired character width, as always, but it is
C interpreted as an angle, in degrees.  The characters of the string
C are initially thought of as being written along the equator of the
C globe and are positioned relative to the point with latitude 0 and
C longitude 0 as determined by the value of the argument CNTR.  The
C variables PANG, PLAT, and PLON, in the common block PCMP04, must have
C been set by the user to determine how the string is to be transformed
C on the surface of the globe.  Three rotations and a call to the EZMAP
C routine MAPTRA are used: the net effect is to move the string to the
C position (PLAT,PLON) in such a way that a vector in the writing
C direction of the string makes an angle of PANG with the local east
C vector.  The reason for doing all this is that it produces little
C distortion in the character string, as viewed on the surface of the
C globe.
C
C In this version of PCMPXY, 'MAP' = 100 selects the identity mapping,
C but a check is made to see whether the point (XINP,YINP) is outside
C the rectangle specified by the values of the variables XVPL, XVPR,
C YVPB, and YVPT, in the common block PCSTCM; if so, the value 1.E12
C is returned for both XOTP and YOTP.  This mapping is used by the
C package STITLE to clip scrolled titles at the edges of the viewport.
C In future versions of PCMPXY, this situation may change; users who
C grab a copy of the routine and modify it should beware of this.
C
C When IMAP is positive, XINP and YINP are the coordinates of a point
C to be mapped to the user coordinate system; the coordinates in the
C user system are to be returned in the variables XOTP and YOTP.  When
C IMAP is negated, the inverse mapping is requested:  (XINP,YINP) is a
C point in the current user coordinate system; (XOTP,YOTP) is returned
C and is the point which would be carried into (XINP,YINP) by the
C mapping numbered ABS(IMAP).
C
C A call with IMAP = 0 should have the form
C
C       CALL PCMPXY (0,REAL(IMAP),RFLG,DUM1,DUM2)
C
C and will return information in RFLG about the mapping numbered IMAP,
C as follows:
C
C   RFLG       forward mapping defined?     inverse mapping defined?
C   ----       ------------------------     ------------------------
C    0.                 no                            no
C    1.                yes                            no
C    2.                 no                           yes
C    3.                yes                           yes
C
C Declare the common block in which TDPACK has put the current window
C limits.
C
        COMMON /TDCOM1/ IH,IT,XM,YM,ZM,XO,YO,ZO,XT,YT,ZT,OE,XE,YE,ZE
        COMMON /TDCOM1/ A1,B1,C1,D1,E1,A2,B2,C2,D2,E2,A3,B3,C3,D3,E3
        COMMON /TDCOM1/ IS,FV,VL,VR,VB,VT,WL,WR,WB,WT
        SAVE   /TDCOM1/
C
C Declare the common block in which STITLE communicates to PLOTCHAR the
C coordinates of the viewport outside which scrolled titles are to be
C clipped.
C
        COMMON /PCSTCM/ XVPL,XVPR,YVPB,YVPT
        SAVE   /PCSTCM/
C
C Declare the common block in which, when IMAP =4, we may find the
C angle at which a point label is to be written on the globe and the
C latitude and longitude of the point being labelled.
C
        COMMON /PCMP04/ PANG,PLAT,PLON
        SAVE   /PCMP04/
C
C Define multiplicative constants to convert from degrees to radians
C and from radians to degrees.
C
        DATA DTOR / .017453292519943 /
        DATA RTOD / 57.2957795130823 /
C
C Sort out the different cases.  When IMAP = 0, PCMPXY is being asked
C what capabilities it has.  The IF tests distinguishing various values
C of XINP from one another make no difference here, since inverses are
C available for all mappings done by this version; it is included as a
C model for those who pick up the routine and modify it (possibly by
C the addition of a mapping for which the inverse is not available).
C
        IF (IMAP.EQ.0) THEN
C
          IF (INT(XINP).EQ.1) THEN
            YINP=3.
          ELSE IF (INT(XINP).EQ.2) THEN
            YINP=3.
          ELSE IF (INT(XINP).EQ.3) THEN
            YINP=3.
          ELSE IF (INT(XINP).EQ.4) THEN
            YINP=1.
          ELSE IF (INT(XINP).EQ.100) THEN
            YINP=3.
          ELSE
            YINP=3.
          END IF
C
C Handle the EZMAP case ...
C
        ELSE IF (ABS(IMAP).EQ.1) THEN
C
          IF (IMAP.GT.0) THEN
            CALL MAPTRA (YINP,XINP,XOTP,YOTP)
            IF (ICFELL('PCMPXY',1).NE.0) RETURN
          ELSE
            CALL MAPTRI (XINP,YINP,YOTP,XOTP)
            IF (ICFELL('PCMPXY',2).NE.0) RETURN
          END IF
C
C ... the polar coordinate case ...
C
        ELSE IF (ABS(IMAP).EQ.2) THEN
C
          IF (IMAP.GT.0) THEN
            XOTP=XINP*COS(DTOR*YINP)
            YOTP=XINP*SIN(DTOR*YINP)
          ELSE
            XOTP=SQRT(XINP*XINP+YINP*YINP)
            YOTP=RTOD*ATAN2(YINP,XINP)
          END IF
C
C ... the perspective case ...
C
        ELSE IF (ABS(IMAP).EQ.3) THEN
C
          IF (IMAP.GT.0) THEN
            CALL TDPRPA (XINP,YINP,XOTP,YOTP)
            IF (XOTP.LT.WL.OR.XOTP.GT.WR.OR.
     +          YOTP.LT.WB.OR.YOTP.GT.WT) THEN
              XOTP=1.E12
              YOTP=1.E12
            END IF
          ELSE
            IF (XINP.LT.WL.OR.XINP.GT.WR.OR.
     +          YINP.LT.WB.OR.YINP.LT.WT) THEN
              XOTP=1.E12
              YOTP=1.E12
            ELSE
              CALL TDPRPI (XINP,YINP,XOTP,YOTP)
            END IF
          END IF
C
C ... a special use of EZMAP which avoids distorting the characters ...
C
        ELSE IF (ABS(IMAP).EQ.4) THEN
C
          UCRD=COS(DTOR*YINP)*COS(DTOR*XINP)
          VCRD=COS(DTOR*YINP)*SIN(DTOR*XINP)
          WCRD=SIN(DTOR*YINP)
          CALL NGRITD (1, PANG,UCRD,VCRD,WCRD)
          CALL NGRITD (2,-PLAT,UCRD,VCRD,WCRD)
          CALL NGRITD (3, PLON,UCRD,VCRD,WCRD)
          RLAT=RTOD*ASIN(WCRD)
          RLON=RTOD*ATAN2(VCRD,UCRD)
          CALL MAPTRA (RLAT,RLON,XOTP,YOTP)
          IF (ICFELL('PCMPXY',3).NE.0) RETURN
C
C ... the special STITLE case ...
C
        ELSE IF (ABS(IMAP).EQ.100) THEN
C
          IF (XINP.LT.XVPL.OR.XINP.GT.XVPR.OR.
     +        YINP.LT.YVPB.OR.YINP.GT.YVPT) THEN
            XOTP=1.E12
            YOTP=1.E12
          ELSE
            XOTP=XINP
            YOTP=YINP
          END IF
C
C ... and the identity case.
C
        ELSE
C
          XOTP=XINP
          YOTP=YINP
C
        END IF
C
C Done.
C
        RETURN
C
      END

File: ncarg2d/man/plotchar/pcmpxy.m
.TH PCMPXY 3NCARG "August 1993" UNIX "NCAR GRAPHICS"
.na
.nh
.SH NAME
PCMPXY - 
This routine is normally not called directly by the user
(though it can be). It is called by each of the routines PCHIQU, PCMEQU,
and PCLOQU when the user has set the mapping flag \'MA\' nonzero to
request mapping of characters from one X/Y coordinate
system to another.
.SH SYNOPSIS
CALL PCMPXY (IMAP,XINP,YINP,XOTP,YOTP)
.SH C-BINDING SYNOPSIS
#include <ncarg/ncargC.h>
.br
void c_pcmpxy (int imap, float xinp, float yinp, float *xotp, float *yotp)
.SH DESCRIPTION 
.IP IMAP 12
(an input expression of type INTEGER) specifies
which mapping is to be done. Whenever Plotchar routines
call PCMPXY, IMAP = \'MA\'. The default version of PCMPXY
recognizes the following values of IMAP:
.RS
.IP "IMAP = 0"
.RS
.IP "Type of mapping" 3
None. When IMAP is zero, the routine PCMPXY is being asked for information
about its capabilities with respect to a particular mapping.
.IP XINP 3
Input: the real equivalent of a possible value of IMAP about which information
is desired.
.IP YINP 3
Output (real): 0. says that the mapping specified by XINP is not available, 1.
that it is available, 2. that its inverse is available, and 3. that both it and
its inverse are available.
.RE
.IP "IMAP = 1"
.RS
.IP "Type of mapping" 3
Ezmap projection.
.IP XINP 3
Longitude, in degrees.
.IP YINP 3
Latitude, in degrees.
.RE
.IP "IMAP = 2"
.RS
.IP "Type of mapping" 3
Rho/theta mapping.
.IP XINP 3
Rho, in user units.
.IP YINP 3
Theta, in degrees.
.RE
.IP "IMAP = 3"
.RS
.IP "Type of mapping" 3
Projection from an arbitrary parallelogram in 3-space to an image plane. This is
implemented using a package which is currently under development and is not yet
generally available.
.IP XINP 3
A multiplier for a unit vector along one edge of the parallelogram.
.IP YINP 3
A multiplier for a unit vector along an intersecting edge of the parallelogram.
.RE
.IP "IMAP = 4"
.RS
.IP "Type of mapping" 3
A version of the EZMAP projection that is intended to be used to label specific
points on the surface of the globe.  When 'MA' = 4, use "0." for each of the
arguments XPOS, YPOS, and ANGD in calls to PCHIQU and PCMEQU.  Initially, each
label is assumed to be placed at the point on the globe with latitude zero and
longitude zero and to be written along the equator.  Three rotations are then
performed to move the label to the point (PLAT,PLON) in such a way that it
makes an angle PANG with the local eastward-pointing vector.  Finally, the
EZMAP routine MAPTRA is called to project the label as specified by the current
state of EZMAP.  The desired values of PANG, PLAT, and PLON (in degrees) are
passed to PCMPXY in the labelled common block PCMP04, which contains just those
three variables, in that order.  Use the ncargex command to see the following
relevant example: cpex10.
.IP XINP 3
Longitude, in degrees.
.IP YINP 3
Latitude, in degrees.
.RE
.IP "IMAP = 100"
.RS
.IP "Type of mapping" 3
The identity mapping, but returns 1.E12 outside a specified viewport. Used
by the utility Scrolled_title to do clipping at the edges of the viewport.
.IP XINP 3
X coordinate, in the fractional system.
.IP YINP 3
Y coordinate, in the fractional system.
.RE
.IP "other > 0"
.RS
.IP "Type of mapping" 3
The identity mapping.
.IP XINP 3
Any real value.
.IP YINP 3
Any real value.
.RE
.IP "other < 0"
.RS
.IP "Type of mapping" 3
An inverse mapping.
.IP XINP 3
A mapped X coordinate.
.IP YINP 3
A mapped Y coordinate.
.RE
.RE
.IP "" 12
Note: Values of IMAP less than or equal to zero are not
currently used by Plotchar in calling PCMPXY. The scheme
described is the same as that used for the routine CPMPXY,
in the NCAR Graphics package Conpack. It is possible that,
in the future, values of IMAP less than or equal to zero
will occur in calls to PCMPXY, so it is best to be prepared.
.IP "XINP and YINP" 12 
(input expressions of type REAL)
specify the X and Y coordinates of a point to be mapped.
.sp
Note: When IMAP is less than or equal to zero (which cannot
currently happen, but may someday be possible), XINP and
YINP are used differently, as shown above.
.IP "XOTP and YOTP" 
(output variables of type REAL) 
in which
PCMPXY returns the mapped X and Y coordinates of the point.
If the point is not visible under the mapping selected (as,
for example, when the orthographic projection of Ezmap is
being used and the point to be mapped is on the other side
of the globe), PCMPXY must return, as a signal that this is
the case, whatever special nonzero value has been given to
the internal parameter \'OR\', so that the calling routine
can take appropriate action (to properly clip the character
being drawn, for example).
.sp
Note: When IMAP is less than or equal to zero (which cannot
currently happen, but may someday be possible), XOTP and
YOTP are used differently, as shown in the table above.
.SH C-BINDING DESCRIPTION
The C-binding argument descriptions are the same as the FORTRAN 
argument descriptions.
.SH USAGE
This routine is normally not called directly by the user
(though it can be). It is called by each of the routines PCHIQU, PCMEQU, and
PCLOQU when the user has set the mapping flag \'MA\' nonzero to
request mapping of characters from one X/Y coordinate
system to another. A call to PCMPXY has the following form:
.sp
.in +5
CALL PCMPXY (IMAP,XINP,YINP,XOTP,YOTP)
.in -5
.sp
The default version of PCMPXY does several useful mappings;
a user version may be supplied to do others.
.SH ACCESS
To use PCMPXY or c_pcmpxy, load the NCAR Graphics libraries ncarg, ncarg_gks,
and ncarg_c, preferably in that order.  
.SH SEE ALSO
Online:
plotchar,
pcdlsc,
pcgetc,
pcgeti,
pcgetr,
pchiqu,
pcloqu,
pcmequ,
pcpnwi,
pcrset,
pcsetc,
pcseti,
pcsetr,
ncarg_cbind.
.sp
Hardcopy:
NCAR Graphics Fundamentals, UNIX Version
.SH COPYRIGHT
Copyright (C) 1987-2009
.br
University Corporation for Atmospheric Research
.br
The use of this Software is governed by a License Agreement.
